x-pack/plugins/security_solution/public/flyout/shared/hooks/use_on_close.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback } from 'react';
+import { useWhichFlyout } from '../../document_details/shared/hooks/use_which_flyout';
+import { Flyouts } from '../../document_details/shared/constants/flyouts';
+import { SECURITY_SOLUTION_ON_CLOSE_EVENT, TIMELINE_ON_CLOSE_EVENT } from '../..';
+
+export interface UseOnCloseParams {
+  /**
+   * Function to call when the event is dispatched
+   */
+  callback: (id: string) => void;
+}
+
+/**
+ * Hook to abstract the logic of listening to the onClose event for the Security Solution application.
+ * The kbn-expandable-flyout package provides the onClose callback, but has there are only 2 instances of the expandable flyout in Security Solution (normal and timeline)
+ * we need a way to propagate the onClose event to all other components.
+ * 2 event names are available, we pick the correct one depending on which flyout is open (if the timeline flyout is open, it is always on top, so we choose that one).
+ */
+export const useOnClose = ({ callback }: UseOnCloseParams): void => {
+  const flyout = useWhichFlyout();
+
+  const eventName =
+    flyout === Flyouts.securitySolution
+      ? SECURITY_SOLUTION_ON_CLOSE_EVENT
+      : TIMELINE_ON_CLOSE_EVENT;
+
+  const eventHandler = useCallback(
+    (e: CustomEventInit) => {
+      callback(e.detail);
+      window.removeEventListener(eventName, eventHandler);
+    },
+    [callback, eventName]
+  );
+
+  window.addEventListener(eventName, eventHandler);
```

Comment:
```suggestion
useEffect(()=>{
  window.addEventListener(eventName, eventHandler);
  
  return () => window.removeEventListener(eventName, eventHandler);
  },[])
  
  
  This change will result in no-memory leak and you will not need to `window.removeEventListener(eventName, eventHandler);` on line 37 explicitly. Since flyout is closing so the event listener will also go away. 
```

Reply:
hmmm I'm also a bit confused here. We only add the listener where the `useOnClose` is called, so unless the flyout is not closed, the listener will be removed eventually. Can you elaborate a bit?

Reply:
So my main point was to define adding event listener within `useEffect`. adding event listener creates without `useEffect` creates following problem.


1. Component A calls hook `useOnClose`.
    - a event listener is created.
2a. Component A is updated/re-renders for some reason.
    - a new instance of event listener is created, while previous one is already there.
    - `Solution` -> No new listener should be created.
2b Alternatively, the callback being passed to `useOnClose` in Component A is updated.
    - a new listener is created with the new callback while the previous one is not removed.
    - `Solution` -> Old listener with old callback should be removed, wthout executing the callback and new listener with `new` callback should added.

Reply:
>You're rising a good point, and I remember now having similar thoughts when I was working on it weeks ago. 

Yes. I think in general, we should be careful against adding any statements in hook/components without react hooks such as `useCallback` / `useEffect` etc.

>What if multiple components use that hook with each their callback. I should test that scenario, as I have a feeling the current approach is not going to support that, which isn't good

From the looks of it.. I think it might be okay becuase it will create a new instance of `listener` with different callbacks for all those components..

Reply:
thanks for the explanation. The change has been implemented in [this commit](https://github.com/elastic/kibana/pull/188196/commits/949fd4e0718a7561e584563fc2d06cda01cbf607)